src/scirpy/ir_dist/_clonotype_neighbors.py

```diff
@@ -203,125 +231,205 @@ def _make_chain_count(self, clonotype_table) -> dict[str, int]:
-            )
-        else:
-            logging.info(
-                "NB: Computation happens in chunks. The progressbar only advances " "when a chunk has finished. "
-            )  # type: ignore
-
-            dist_rows = process_map(
-                self._dist_for_clonotype,
-                range(n_clonotypes),
-                max_workers=_get_usable_cpus(self.n_jobs),
-                chunksize=2000,
-                tqdm_class=tqdm,
-            )
+        dist = self._dist_for_clonotype(clonotype_ids)
 
-        dist = sp.vstack(list(dist_rows))
         dist.eliminate_zeros()
         logging.hint("Done computing clonotype x clonotype distances. ", time=start)
         return dist  # type: ignore
 
-    def _dist_for_clonotype(self, ct_id: int) -> sp.csr_matrix:
-        """Compute neighboring clonotypes for a given clonotype.
-
-        Or operations use the min dist of two matching entries.
-        And operations use the max dist of two matching entries.
-
-        The motivation for using the max instead of sum/average is
-        that our hypotheis is that a receptor recognizes the same antigen if it
-        has a sequence dist < threshold. If we require both receptors to
-        match ("and"), the higher one should count.
-        """
-        # Lookup distances for current row
-        tmp_clonotypes = self.clonotypes2 if self.clonotypes2 is not None else self.clonotypes
-        lookup = {}  # CDR3 distances
-        lookup_v = {}  # V-gene distances
+    def _dist_for_clonotype(self, ct_ids: np.ndarray[int]) -> sp.csr_matrix:
+        lookup = {}
+        chain_ids = [(1, 1)] if self.dual_ir == "primary_only" else [(1, 1), (2, 2), (1, 2), (2, 1)]
         for tmp_arm in self._receptor_arm_cols:
-            chain_ids = [(1, 1)] if self.dual_ir == "primary_only" else [(1, 1), (2, 2), (1, 2), (2, 1)]
             for c1, c2 in chain_ids:
                 lookup[(tmp_arm, c1, c2)] = self.neighbor_finder.lookup(
-                    ct_id,
+                    ct_ids,
                     f"{tmp_arm}_{c1}",
                     f"{tmp_arm}_{c2}",
                 )
-                if self.same_v_gene:
-                    lookup_v[(tmp_arm, c1, c2)] = self.neighbor_finder.lookup(
-                        ct_id,
-                        f"{tmp_arm}_{c1}_v_call",
-                        f"{tmp_arm}_{c2}_v_call",
-                    )
+        id_len = len(ct_ids)
+
+        first_value = next(iter(lookup.values()))
+        has_distance_table = sp.csr_matrix((id_len, first_value.shape[1]))
+        for value in lookup.values():
+            has_distance_table += value
+
+        has_distance_mask = has_distance_table
+        has_distance_mask.data = np.ones_like(has_distance_mask.data)
+
+        def csr_min(a, b):
+            max_value_a = np.max(a.data, initial=0)
+            max_value_b = np.max(b.data, initial=0)
+            max_value = np.int16(np.max([max_value_a, max_value_b]) + 1)
```

Comment:
Why the `np.int16` here?

Reply:
```
        def csr_min(a, b):
            max_value_a = np.max(a.data, initial=0)
            max_value_b = np.max(b.data, initial=0)
            max_value = np.int16(np.max([max_value_a, max_value_b]) + 1)
            min_mat_a = sp.csr_matrix((a.data.astype(np.int16), a.indices, a.indptr), shape=a.shape)
            min_mat_a.data -= max_value
            min_mat_b = sp.csr_matrix((b.data.astype(np.int16), b.indices, b.indptr), shape=b.shape)
            min_mat_b.data -= max_value
            a_smaller_b = min_mat_a < min_mat_b
            return b + (a - b).multiply(a_smaller_b)
```

The idea behind this function is to get the minimum values of 2 csr matrices while ignoring 0 values with built-in functions. The trick here is to shift the data arrays by the maximum value, such that you get negative values. Then you can use the built in function "<" and ignore the 0 values. int16 is needed to allow negative values and cover at least all values that are are possible with uint8.

Reply:
Ok, thanks for explaining! Would you mind adding this as a comment in the code? 

Should the distance matrices not be of uint8 type, would this fail, or lead to incorrect results? If it doesn't fail explicitly, I'd add an assertion for the data type.

Reply:
> Ok, thanks for explaining! Would you mind adding this as a comment in the code?

Yes, I will add more documentation to all of the functions!

> Should the distance matrices not be of uint8 type, would this fail, or lead to incorrect results? If it doesn't fail explicitly, I'd add an assertion for the data type.

I will have to take a look at the datatypes in general. Probably i should already restrict the data type to uint8 at the distance matrix computation. Do you think anyone would want to set a cutoff > 255 (uint max)? That's probably too high even for the tcrdist.

Reply:
> Probably i should already restrict the data type to uint8 at the distance matrix computation

yes, but it wouldn't prevent anyone from building their own distance calculator class or just overwriting the distance values in `AnnData.uns` for whatever reason. So we still need checks here. 

> Do you think anyone would want to set a cutoff > 255 (uint max)? That's probably too high even for the tcrdist.

I don't think this would make sense for any of the metrics we use.